File: cloud_functions/functions/src/models/conversation.ts
export class Conversation {
  users: ConversationUser[] = [];
  userIds: string[] = [];
  userIdsHash?: string = '';
  latestMessage?: string | undefined;
  latestMessageId?: string | undefined;
  latestMessageTimestamp?: FirebaseFirestore.Timestamp | undefined;
  latestMessageSentBy?: string | undefined;
  seen: { [userId: string]: string } = {};
}

export class ConversationWithId extends Conversation {
  uid: string = '';

  static fromBase(uid: string, user: Conversation): ConversationWithId {
    return { uid, ...user };
  }
}

export class ConversationUser {
  userId: string = '';
  userName: string = '';
  imageUrl: string = '';
}

File: mobile/pa_messenger/lib/models/conversation.dart
import 'package:cloud_firestore/cloud_firestore.dart';

import 'conversationUser.dart';

class Conversation {

  String id = '';

  List<ConversationUser> users;
  List<String> userIds;
  String latestMessage;
  String latestMessageId;
  Timestamp latestMessageTimestamp;
  String latestMessageSentBy;

  Map<String, dynamic> seen; // { userId: messageId }

  Conversation({
    this.id,
    this.users,
    this.userIds,
    this.latestMessage,
    this.latestMessageId,
    this.latestMessageTimestamp,
    this.latestMessageSentBy,
    this.seen,
  });

  factory Conversation.fromMap(Map<String, dynamic> map) {
    return Conversation(
      id: map['id'],
      users: ConversationUser.fromMapList(map['users'] ?? []),
      seen: map['seen'] ?? {},
      userIds: List<String>.from(map['userIds'] ?? []),
      latestMessage: map['latestMessage'],
      latestMessageId: map['latestMessageId'] ?? null,
      latestMessageTimestamp: map['latestMessageTimestamp'],
      latestMessageSentBy: map['latestMessageSentBy'],
    );
  }

  static List<Conversation> fromMapList(List<Map<String, dynamic>> list) {
    return list.map((x) => Conversation.fromMap(x)).toList();
  }
}
